### PR TITLE
Mule 16308: Local authentication doesn't prevail when there're multiple "WWW-Authenticate" headers.

### PR DESCRIPTION
**Pull request by emariotti3**

### CODE REVIEW

**Comment by emariotti3** [comment]
--validate

File: pom.xml [hunk 1]
```diff
@@ -184,6 +184,12 @@
             <version>1.1.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (fsgonz) [inline pom.xml]: Why is this needed?

File: src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    return responseBuilder.build();
+  }
+
+  public HttpRequestOptions getOptionsDigest() {
+    HttpAuthentication authentication =
+        HttpAuthentication.digest("user", "password").preemptive(false).build();
+    return HttpRequestOptions.builder().authentication(authentication).build();
+  }
+
+  public HttpRequestOptions getOptionsNTLM() {
```
Comment (fsgonz) [inline src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java]: getOptionsNTLM -> getOptionsNtlm according to what @afelisatti noticed.

File: src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      choseNtlm = true;
+    } else {
+      responseBuilder.statusCode(UNAUTHORIZED.getStatusCode());
+    }
+    return responseBuilder.build();
+  }
+
+  public HttpRequestOptions getOptionsDigest() {
+    HttpAuthentication authentication =
+        HttpAuthentication.digest("user", "password").preemptive(false).build();
```
Comment (fsgonz) [inline src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java]: try to use constants even in these cases.

File: src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  @Override
+  protected HttpResponse setUpHttpResponse(HttpRequest request) {
+    HttpResponseBuilder responseBuilder = HttpResponse.builder();
+    String authHeader = request.getHeaderValue(AUTHORIZATION);
+
+    if (StringUtils.isEmpty(authHeader)) {
+      responseBuilder.statusCode(UNAUTHORIZED.getStatusCode())
+          .addHeader(WWW_AUTHENTICATE, "Basic")
```
Comment (fsgonz) [inline src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java]: Isn't there a constant for "Basic" authentication.

File: src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  public HttpRequestMultipleAuthenticationMethodsSupportedTestCase(String serviceToLoad) {
+    super(serviceToLoad);
+  }
+
+  @Override
+  protected HttpResponse setUpHttpResponse(HttpRequest request) {
+    HttpResponseBuilder responseBuilder = HttpResponse.builder();
+    String authHeader = request.getHeaderValue(AUTHORIZATION);
+    String authType = getRequestAuthType(authHeader);
+    chosenAuthMethod = "";
```
Comment (afelisatti) [inline src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java]: Why do you need this extra variable? Wouldn't `authType` suffice?

File: src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+  private class ResponseSuccessProbe implements Probe {
+
+    @Override
+    public boolean isSatisfied() {
+      return responseReference.get().getStatusCode() == OK.getStatusCode();
+    }
+
+    @Override
+    public String describeFailure() {
+      return "Authorization method chosen should be Basic!";
```
Comment (afelisatti) [inline src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java]: You'll be saying it needs to be `Basic` on all tests. You need to parameterize this and use the `authMethod` data to instantiate the `Probe`.

File: src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  @After
+  public void stopClient() {
+    if (client != null) {
+      client.stop();
+    }
+  }
+
+  @Test
+  public void onMultipleAuthenticationMethodsSupportedLocallySetMethodIsChosenDigest() {
+    sendRequestAssertCorrectAuthMethodChosen(HttpAuthentication.digest(TEST_USER, TEST_PASS), DIGEST);
```
Comment (afelisatti) [inline src/test/java/org/mule/service/http/impl/service/server/grizzly/HttpRequestMultipleAuthenticationMethodsSupportedTestCase.java]: You are still repeating the user and pass set up. Use the HttpAuthentication `builder()` method and the builder's `type()` method.
